File: src/visual-graph/builder/set-predecessor-of.test.ts
import { describe, expect, test } from "vitest";

import { asScopeId } from "../../ir/serialized/scope-id.js";
import type { SerializedScope } from "../../ir/serialized/serialized-scope.js";
import { setPredecessorOf } from "./set-predecessor-of.js";
import { baseScope } from "./testing/make-scope.js";
import { baseWriteOp } from "./testing/make-write-op.js";

describe("setPredecessorOf", () => {
  test("returns variable node id when there are no prior ops", () => {
    const scope = { ...baseScope(), id: asScopeId("s") };
    const map = new Map<string, SerializedScope>([[scope.id, scope]]);
    const op = { ...baseWriteOp(), refId: "r1", scopeId: "s" };
    expect(setPredecessorOf(op, [op], map)).toEqual("n_v");
  });

  test("returns prior op's writeOp node id when prior op is in the same scope", () => {
    const scope = { ...baseScope(), id: asScopeId("s") };
    const map = new Map<string, SerializedScope>([[scope.id, scope]]);
    const prev = { ...baseWriteOp(), refId: "rPrev", scopeId: "s" };
    const op = { ...baseWriteOp(), refId: "rCur", scopeId: "s" };
    expect(setPredecessorOf(op, [prev, op], map)).toEqual("wr_rPrev");
  });

  test("returns prior op's writeOp node id when prior op is in an ancestor scope", () => {
    const root = { ...baseScope(), id: asScopeId("root") };
    const child = {
      ...baseScope(),
      id: asScopeId("child"),
      upper: asScopeId("root"),
    };
    const map = new Map<string, SerializedScope>([
      [root.id, root],
      [child.id, child],
    ]);
    const prev = { ...baseWriteOp(), refId: "rPrev", scopeId: "root" };
    const op = { ...baseWriteOp(), refId: "rCur", scopeId: "child" };
    expect(setPredecessorOf(op, [prev, op], map)).toEqual("wr_rPrev");
  });

  test("returns variable node id when prior op is in a sibling scope", () => {
    const root = { ...baseScope(), id: asScopeId("root") };
    const sibA = {
      ...baseScope(),
      id: asScopeId("a"),
      upper: asScopeId("root"),
    };
    const sibB = {
      ...baseScope(),
      id: asScopeId("b"),
      upper: asScopeId("root"),
    };
    const map = new Map<string, SerializedScope>([
      [root.id, root],
      [sibA.id, sibA],
      [sibB.id, sibB],
    ]);
    const prev = { ...baseWriteOp(), refId: "rPrev", scopeId: "a" };
    const op = { ...baseWriteOp(), refId: "rCur", scopeId: "b" };
    expect(setPredecessorOf(op, [prev, op], map)).toEqual("n_v");
  });

  test("returns variable node id when op is missing from ops list", () => {
    const scope = { ...baseScope(), id: asScopeId("s") };
    const map = new Map<string, SerializedScope>([[scope.id, scope]]);
    const op = { ...baseWriteOp(), refId: "rCur", scopeId: "s" };
    expect(setPredecessorOf(op, [], map)).toEqual("n_v");
  });
});

File: crates/unsnarl-visual-graph/src/builder/set_predecessor_of_test.rs
//! Sibling tests for [`set_predecessor_of`]. Cases mirror
//! `ts/src/visual-graph/builder/set-predecessor-of.test.ts`.

use std::collections::HashMap;

use unsnarl_ir::serialized::SerializedScope;

use super::set_predecessor_of;
use crate::builder::testing::{base_serialized_scope, base_write_op, scope_id};
use crate::builder::write_op::WriteOp;

fn write_op_with(ref_id: &str, scope_id: &str) -> WriteOp {
    let mut op = base_write_op();
    op.ref_id = ref_id.to_string();
    op.scope_id = scope_id.to_string();
    op
}

fn scope_map(scopes: &[SerializedScope]) -> HashMap<&str, &SerializedScope> {
    scopes.iter().map(|s| (s.id.value(), s)).collect()
}

#[test]
fn returns_variable_node_id_when_there_are_no_prior_ops() {
    let scopes = [base_serialized_scope("s")];
    let map = scope_map(&scopes);
    let op = write_op_with("r1", "s");
    assert_eq!(
        set_predecessor_of(&op, std::slice::from_ref(&op), &map),
        "n_v"
    );
}

#[test]
fn returns_prior_write_node_id_when_prior_in_the_same_scope() {
    let scopes = [base_serialized_scope("s")];
    let map = scope_map(&scopes);
    let prev = write_op_with("rPrev", "s");
    let op = write_op_with("rCur", "s");
    assert_eq!(
        set_predecessor_of(&op, &[prev, op.clone()], &map),
        "wr_rPrev"
    );
}

#[test]
fn returns_prior_write_node_id_when_prior_in_an_ancestor_scope() {
    let root = base_serialized_scope("root");
    let mut child = base_serialized_scope("child");
    child.upper = Some(scope_id("root"));
    let scopes = [root, child];
    let map = scope_map(&scopes);
    let prev = write_op_with("rPrev", "root");
    let op = write_op_with("rCur", "child");
    assert_eq!(
        set_predecessor_of(&op, &[prev, op.clone()], &map),
        "wr_rPrev"
    );
}

#[test]
fn returns_variable_node_id_when_prior_in_a_sibling_scope() {
    let root = base_serialized_scope("root");
    let mut a = base_serialized_scope("a");
    a.upper = Some(scope_id("root"));
    let mut b = base_serialized_scope("b");
    b.upper = Some(scope_id("root"));
    let scopes = [root, a, b];
    let map = scope_map(&scopes);
    let prev = write_op_with("rPrev", "a");
    let op = write_op_with("rCur", "b");
    assert_eq!(set_predecessor_of(&op, &[prev, op.clone()], &map), "n_v");
}

#[test]
fn returns_variable_node_id_when_op_is_missing_from_ops_list() {
    let scopes = [base_serialized_scope("s")];
    let map = scope_map(&scopes);
    let op = write_op_with("rCur", "s");
    assert_eq!(set_predecessor_of(&op, &[], &map), "n_v");
}
